### ISSUE TITLE
BIG PICTURE IDEA - Upload/Import multiple clients at once from spreadsheet/csv

### ISSUE THREAD

**Issue by dan-langridge** [NorwoodAndBrixtonFoodbank/nbf-website#274]
### Describe the solution / feature you'd like

At Christmas for example, we want to bulk add clients into the app to save time.
It would be cool if there was a way for an admin to receive a spreadsheet from the client upload team to mass import multiple clients at once.

For example, a Hostel gives us the names of 40 clients who will all receive a parcel at Christmas. Instead of manually adding one at a time using the in-built form, a bulk upload would be much faster. 

Would require someone with good spreadsheet knowledge to check all the data is correct and in the right format otherwise could upload wrong or break database. Admin only for sure.

**Comment by Stuart-Knowles-SW** [comment]
Hi Dan,

For this, I was thinking we could create a spreadsheet template for you that could be provided to the client upload team for them to add all the new clients. That's then sent back to the admin, who can upload it through the app.

We can then parse that spreadsheet and automatically validate all the data before actually putting it in the database, which should save you having to check it by hand. We can return separate lists of clients that uploaded successfully and those that had problems, so the problems can be fixed and those clients re-uploaded. It should be entirely safe against malformed or bad data, and allow you to just glance at the spreadsheet before uploading, then fix problems only when they arise.

How does that sound?

**Comment by dan-langridge** [comment]
Sounds great, thanks!On 4 Jul 2024, at 09:56, Stuart Knowles ***@***.***> wrote:﻿
Hi Dan,
For this, I was thinking we could create a spreadsheet template for you that could be provided to the client upload team for them to add all the new clients. That's then sent back to the admin, who can upload it through the app.
We can then parse that spreadsheet and automatically validate all the data before actually putting it in the database, which should save you having to check it by hand. We can return separate lists of clients that uploaded successfully and those that had problems, so the problems can be fixed and those clients re-uploaded. It should be entirely safe against malformed or bad data, and allow you to just glance at the spreadsheet before uploading, then fix problems only when they arise.
How does that sound?

**Comment by stonelink** [comment]
Thanks @Stuart-Knowles-SW and @dan-langridge :) 

A likely part of this scope that Dan and I discussed on a call but I didn't track on the ticket was a 'Merge Clients' function. That is, it's possible/likely that there will already be a Client record in the system for at least some of the clients that are being imported from the spreadsheet.

We don't currently have a Client ID that Trussell Trust would recognise (only voucher IDs, which are attached to Parcels). That means that we have to recognise duplicates by comparing records - presumably Name, Phone & Address. We need to think about:
- What if other fields don't match, like family details or dietary requirements?
- What if someone moves house, or their phone number is formatted differently?

Giving Admins a Merge Clients function that gives fine control would be a separate ticket, but can we think about:
- Is there a reasonable minimal solution for definite duplicates?
- How might a fully-functional solution work? (We might want to implement that before Christmas)


And one other question for @dan-langridge: is there a Client ID that we can/should get from Trussell Trust that would let us know for sure which Client is being referred to?

**Comment by Stuart-Knowles-SW** [comment]
Thanks @stonelink, that's good to know. A think a reasonable approach here is to give the admin three things instead of two:

- Clients who were uploaded successfully
- Clients with malformed data which needs to be fixed for re-upload
- Clients with potential duplicates, where we show both the existing entry and the new one for the admin to compare

With the last group, I agree that fine-grained automated control is out of scope for this, so for now could we just show that info to the admin in as nice a format as possible and let them handle the resolution manually?

**Comment by dan-langridge** [comment]
I'm looking into whether there is a unique client ID generated by Trussel Trust.

**Comment by stonelink** [comment]
Hi @dan-langridge just to clarify this spec - my understanding is that each row of the spreadsheet will contain the details of a Client _and_ their Parcel. (The term "client" in the ticket description is as per the existing system, that makes a new client record for each parcel.)

Please let @Stuart-Knowles-SW and I know if I've misunderstood that.

**Comment by dan-langridge** [comment]
Yes that’s what I would expect to happen, thanks Ricky.So importing from the sheet would create a client AND create 1 parcel for that client at the same time.On 5 Jul 2024, at 14:07, Lincoln Stone ***@***.***> wrote:﻿
Hi @dan-langridge just to clarify this spec - my understanding is that each row of the spreadsheet will contain the details of a Client and their Parcel. (The term "client" in the ticket description is as per the existing system, that makes a new client record for each parcel.)
Please let @Stuart-Knowles-SW and I know if I've misunderstood that.

**Comment by stonelink** [comment]
Moving this back to `In Progress` because this is the Batch Creation function, covered by Jira tickets, which is incomplete
